### PR TITLE
[BEAM-6186] Finish moving optimization phases.

### PR DESCRIPTION
**Pull request by robertwb**
This is mostly a simple refactoring, moving references by closure to
an explicit context argument.  Of note, however, the length-prefix coder
computation was reworked to be cleaner and more consistent.

------------------------

Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

It will help us expedite review of your Pull Request if you tag someone (e.g. `@username`) to look at it.

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_GradleBuild/lastCompletedBuild/) | --- | --- | --- | --- | --- | ---
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_GradleBuild/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink_Gradle/lastCompletedBuild/) [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza_Gradle/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark_Gradle/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_Verify/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/) </br> [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Flink/lastCompletedBuild/) | --- | --- | ---

### CODE REVIEW

**Comment by robertwb** [comment]
Run Portable_Python PreCommit

**Comment by angoenka** [comment]
Run RAT PreCommit

**Comment by robertwb** [comment]
Ping @charlesccychen

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py [hunk 1]
```diff
@@ -157,18 +163,121 @@ def deduplicate_read(self):
...
+
+  @memoize_on_instance
+  def length_prefixed_coder(self, coder_id):
+    if coder_id in self.safe_coders:
+      return coder_id
+    length_prefixed_id, safe_id = self.length_prefixed_and_safe_coder(coder_id)
+    self.safe_coders[length_prefixed_id] = safe_id
+    return length_prefixed_id
+
+  def length_prefixed_and_safe_coder(self, coder_id):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Should we memoize here?

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py [hunk 2]
```diff
@@ -157,18 +163,121 @@ def deduplicate_read(self):
...
+          self.components.coders, coder_id + '_length_prefixed')
+      self.components.coders[new_coder_id].CopyFrom(
+          beam_runner_api_pb2.Coder(
+              spec=beam_runner_api_pb2.SdkFunctionSpec(
+                  spec=beam_runner_api_pb2.FunctionSpec(
+                      urn=common_urns.coders.LENGTH_PREFIX.urn)),
+              component_coder_ids=[coder_id]))
+      return new_coder_id, self.bytes_coder_id
+
+  def length_prefix_pcoll_coders(self, pcoll_id):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Should we memoize here?

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py [hunk 3]
```diff
@@ -353,6 +518,382 @@ def make_stage(base_stage, transform):
...
+      seen.add(stage)
+      for prev in stage.must_follow:
+        process(prev)
+      ordered.append(stage)
+  for stage in stages:
+    process(stage)
+  return ordered
+
+
+def window_pcollection_coders(stages, pipeline_context):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Can we add a docstring here for consistency?

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py [hunk 4]
```diff
@@ -353,6 +518,382 @@ def make_stage(base_stage, transform):
...
+            beam_runner_api_pb2.PTransform(
+                unique_name=transform.unique_name,
+                spec=beam_runner_api_pb2.FunctionSpec(
+                    urn=bundle_processor.DATA_INPUT_URN,
+                    payload=IMPULSE_BUFFER),
+                outputs=transform.outputs))
+    yield stage
+
+
+def inject_timer_pcollections(stages, pipeline_context):
```
Comment (charlesccychen) [inline sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Can we add a docstring here for consistency?

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py (hunk 1 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: It's not generally called directly, but good idea for consistency.

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py (hunk 2 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: I don't think it makes sense here.

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py (hunk 4 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Done.

File: sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py (hunk 3 above)
Comment (robertwb) [reply sdks/python/apache_beam/runners/portability/fn_api_runner_transforms.py]: Done.

**Pull request merged**